### PR TITLE
appveyor: run tests against MSYS2-MinGW

### PR DESCRIPTION
**Pull request by msink**

### CODE REVIEW

**Comment by nirbheek** [comment]
That traceback is really strange:

```
Traceback (most recent call last):
  File "C:/projects/meson/test cases/common/135 generated assembly/copy.py", line 4, in <module>
    import shutil
  File "C:\python34\lib\shutil.py", line 14, in <module>
    import tarfile
  File "C:\python34\lib\tarfile.py", line 49, in <module>
    import copy
  File "C:\projects\meson\test cases\common\135 generated assembly\copy.py", line 6, in <module>
    shutil.copyfile(sys.argv[1], sys.argv[2])
AttributeError: 'module' object has no attribute 'copyfile'
```

**Comment by msink** [comment]
> That traceback is really strange

Fixed, but still 3 tests failed - this bug masked anotner one :)

**Comment by nirbheek** [comment]
Could you please update this again and check what tests pass? I really want to see this merged so we can say with confidence that Meson + MinGW is a first-class citizen. :)

**Comment by msink** [comment]
Yes, but after this weekend

**Comment by msink** [comment]
Updated, failed tests: 9

common\85 internal dependency
common\86 same basename
common\87 declare dep
common\93 private include
common\106 subproject subdir
common\120 subdir subproject
common\127 cpp and asm
common\135 generated assembly
windows\5 resources

**Comment by nirbheek** [comment]
Most of those will be fixed by https://github.com/mesonbuild/meson/pull/1527.

What are actually broken are:

```
common\135 generated assembly
windows\5 resources
```

135 is broken because it assumes intel syntax for the assembly. Just need to port that to GAS. For 64-bit, the Linux version should work as-is. For 32-bit the calling convention is different on Windows so you'd need to port the intel version to GAS.

The windres failure is interesting. Will have to poke at it manually. Can you look into that?

Also, please make it so that the CI fails when MinGW tests fail instead of passing like it does now.

I think we can fix 135, put windres on skip (if you can't figure that out), make passing of these tests compulsory and push this.

**Comment by nirbheek** [comment]
As an aside, the appveyor tests are taking too long to build ever since the addition of vs2017 since we only have one job available on appveyor.

I think we should change the matrix to build VS2015 only on x86, and VS2017 only on x64. That should cut the time by roughly half without losing significant test coverage. I think the VS2015 and VS2017 implementations are identical enough that we don't need to test both on 32-bit and 64-bit.

**Comment by nirbheek** [comment]
I've implemented the matrix change in https://github.com/mesonbuild/meson/pull/1529 to try it out.

**Comment by msink** [comment]
Rebased again, for now fails:

unittest: static_library_overwrite
common\127 cpp and asm (fails in x64 mode)
common\127 cpp and asm
windows\5 resources

**Comment by msink** [comment]
`windows\5 resources` works form directory without spaces in path

**Comment by nirbheek** [comment]
I have a bunch of patches that fix a lot of that. Do you mind if I push it to this PR?

**Comment by msink** [comment]
I do not mind :)

File: mesonbuild/backend/vs2010backend.py [hunk 1]
```diff
@@ -151,7 +151,11 @@ def generate_custom_generator_commands(self, target, parent_node):
                     args = [x.replace("@INPUT@", infilename).replace('@OUTPUT@', sole_output)
                             for x in base_args]
                     args = self.replace_outputs(args, target_private_dir, outfiles_rel)
-                    args = [x.replace("@SOURCE_DIR@", self.environment.get_source_dir()).replace("@BUILD_DIR@", target_private_dir)
+                    args = [x.replace("@SOURCE_DIR@", self.environment.get_source_dir())\
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2010backend.py]: __[flake8]__
* [E502] the backslash is redundant between brackets

File: mesonbuild/backend/vs2010backend.py [hunk 2]
```diff
@@ -151,7 +151,11 @@ def generate_custom_generator_commands(self, target, parent_node):
...
+                             .replace("@BUILD_DIR@", target_private_dir)
+                            for x in args]
+                    args = [x.replace("@SOURCE_ROOT@", self.environment.get_source_dir())\
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2010backend.py]: __[flake8]__
* [E502] the backslash is redundant between brackets

**Comment by nirbheek** [comment]
This might cause lots of failures on Linux. Let's see.

**Comment by nirbheek** [comment]
Yay everything passes. Let's merge this.

File: test cases/windows/5 resources/meson.build [hunk 3]
```diff
@@ -1,9 +1,66 @@
...
-res = win.compile_resources('myres.rc',
-  include_directories : include_directories('inc')
-)
+# MinGW windres has a bug due to which it doesn't parse args with space properly:
+# https://github.com/mesonbuild/meson/pull/1346
+# https://sourceware.org/bugzilla/show_bug.cgi?id=4933
+if meson.get_compiler('c').get_id() == 'gcc'
+  # Construct build_to_src and skip this test if it has spaces
+  # because then the -I flag to windres will also have spaces
+  # and we know the test will fail
```
Comment (jpakkane) [inline test cases/windows/5 resources/meson.build]: Isn't this always the case? The build dir used always has a space and the path to the source tree is `test cases/windows/5 resources/`.

Comment (nirbheek) [inline test cases/windows/5 resources/meson.build]: Not if your builddir is a subdir of the source tree, i.e., when you're running the test manually. The previous commit ensures that relative dirs are used whenever possible to allow this.

**Pull request merged**